### ISSUE TITLE
14.2-DRL: Policy Gradient

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#74]
# Policy Gradient

解 RL拿來做 Optimization 常用的一個演算法
過去上課的錄影對 Policy Gradient 有比較詳細的推導

### 操控Actor 的輸出

讓 Actor它的輸入是 s 的時候 它就要輸出 Action a^ 
也就說 s 是 Actor 的輸入 a^ 就是我們的 Label = Ground Truth (正確答案)
計算你的 Actor它的輸出跟 Ground Truth 之間的 Cross-entropy
你再去 Learn 一個 θ 然後這個 θ 可以讓 Loss 最小
這個是要讓你的 Actor 採取某一個行為的時候的做法

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/962f0c7e-1e03-4bd5-b926-10dd5492641a)

假設你想要讓你的 Actor 不要採取某一個行為的話
你只需要把 Loss 的定義反過來就好
那你就把你就定一個 Loss 叫做負的 Cross-entropy
你去 Minimize 這個 L 就是讓 Cross-entropy 越大越好->讓 a 跟 a^ 的距離越遠越好

只要我們給它適當的 Label 跟適當的 Loss 我們有辦法控制我們的 Actor 做我們想要做的事
所以假設我們要讓我們的 Actor 看到 s 的時候採取 a^ 看到 s' 的時候不要採取 a^' 的話

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/54a26bf6-9fba-4a17-88c1-48076a27eaf3)

我們找一個 θ 去 Minimize Loss 得到 θ⋆
這就跟 Supervised Learning Train 的 Image Classifier 是一模一樣的
但等下我們會看到它跟一般的 Supervised Learning 不一樣在哪裡 

每一個行為就是要執行或不要執行->這是一個 Binary 的問題, 用 ±1 來表示
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/5d886c5f-579c-4bc0-9cef-4cac90487a4d)

但是每一個行為它是有程度的差別的->改成每一個 s 跟 a 的 Pair它有對應的一個分數
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/90a5c8a3-78b5-4244-b3c5-196cfa95118f)

這個分數代表說 我們多希望機器在看到 s1 的時候執行 a1^ 這個行為 

比如說我們期待機器看到 s1 的時候它可以做 a1^ 看到 s3 的時候它可以做 a3^
但是我們期待它看到 s1 的時候 做 a1^ 的這個期待更強烈一點 (1.5>0.5)
那我們希望它在看到 s2 的時候不要做 a2^ (-0.5)

有了這些資訊 你一樣可以定義一個 Loss Function
只是在原來的 Cross-entropy 前面改成乘上 An 這一項

____________

那接下來的難點就是要

怎麼定出這一個 a 呢

怎麼產生這個 s 跟 a 的 Pair 呢 
怎麼知道在 s1 的時候要執行 a1
或在 s2 的時候不要執行 a2 呢

_____________

### 定義 A - version0

先講一個最簡單的 但是其實是不正確的版本(助教Sample Code 的版本)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/bd20a045-3fb5-4e7a-a76a-2b3e9c67d53f)

你先把這個 Actor 想成就是一個隨機的 Actor 
它就是一個隨機的東西 那看到 s1 然後它執行的行為就是亂七八糟的(就是隨機的)
每一個 s 執行的行為 a通通都記錄下來
接下來 我們就去評價每一個 Action 它到底是好還是不好

你不會只把 Actor 跟環境做一個 Episode 通常會做多個 Episode
比如說在助教 Sample Code裡面可能就是跑了 5 個 Episode 然後才蒐集到足夠的資料

A1 就是 r1
A2 就是 r2
A3 就是 r3 
AN 就是 rN 
等同於你就告訴 machine 說
如果我們執行完某一個 Action
得到的 Reward 是正的那這就是一個好的 Action
得到 Reward 是負的 的那這就是一個不-好的 Action

評價完以後 我們就可以拿我們評價的結果 來訓練我們的 Actor

但是每一個行為 其實都會影響互動接下來的發展
也就是說 Actor 在 s1 執行 a1 得到 r1 這個並不是互動的全部
因為 a1 影響了我們接下來會看到 s2 
s2 會影響到接下來會執行 a2 也影響到接下來會產生 r2 
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/e61b9230-f76a-4508-b139-61400f4b513f)

所以每一個行為並不是獨立的 每一個行為都會影響到接下來發生的事情
Reward Delay  = 有時候犧牲短期的利益以換取更長程的目標 

假設如果只有開火的時候它得到的 Reward 才會是正的
Machine 就會學到 它只有瘋狂狂開火才是對的
Version 0 就只會學到瘋狂開火而已

___________

### VERSION1

a1 它有多好 不是在取決於 r1 而是取決於 a1 之後所有發生的事情
我們會把 a1 執行完 a1 以後 所有得到的 Reward r1 r2 r3 到 rN 通通加起來 得到一個數值叫做 G1
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/f65d766a-e5d5-4ed2-89f6-66401d4f5ee4)

但假設這個遊戲非常地長 ->你把 rN 歸功於 a1 好像不太合適

_________

### VERSION2-權重

G' = Cumulated 的 Reward -> 在 r 前面 乘一個 Discount 的 Factor = γ
Discount 的 Factor γ會設一個小於 1 的值
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8c85b2f8-57aa-473a-bf59-49481f67d95c)

用這個方法就可以把離 a1 比較近的那些 Reward給它比較大的權重
離a1比較遠的那些 Reward 給它比較小的權重

### VERSION3-標準化

越早的動作就會累積到越多的分數 越晚的動作累積的分數就越少->需要標準化

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/7058a13b-e59c-41bc-886c-68fea7f5448d)

到遊戲的終局 沒什麼外星人可以殺了 你可能做什麼事對結果影響都不大

好或壞是相對的->需要標準化
最低分就預設是 10 分 那你其實得到 10 分的 Reward 根本算是差的
reward不果怎麼加都會是正的 -> 會變成全部行為都是鼓勵機器去做

一個最簡單的方法就是把所有的 G' 都減掉一個 b = Baseline
目標就是讓 G' 有正有負 
特別高的 G' 讓它是正的
特別低的 G' 讓它是負的

____________

首先你要先 Random 初始化
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/658a079c-da06-4e89-8f7e-898c611148b6)

隨機初始化 Actor 的參數 = θ0

接下來你進入你的 Training Iteration 假設你要跑 T 個 Training Iteration

現在手上有的 Actor 一開始是這個 θ0
一開始很笨 它什麼都不會 它採取的行為都是隨機的 但它會越來越好
你拿你的 Actor 去跟環境做互動 那你就得到一大堆的 s 跟 a
你就得到一大堆的 s 跟 a 就把它互動的過程記錄下來 得到這些 s 跟 a

那接下來你就要進行評價 你用 A1 到 AN 來決定說 這些 Action 到底是好還是不好

拿你的 Actor 去跟環境做互動->蒐集->評價
你最需要把它改動的地方就是在評價這個過程裡面
助教程式這個 A 就直接設成 Immediate Reward
要改這一段 你才有可能得到好的結果

一般的 training (到目前為止的) Training Data Collection 都是在 For 迴圈之外
但在 RL 裡面蒐集資料是在 For 迴圈裡面
假設這個 For 迴圈 你打算跑 400 次那你就得蒐集資料 400 次

你觀察了某一個 Actor它在每一個 State 執行的 Action
然後接下來你給予一個評價 
但要用什麼評價 要用哪一個版本這個是你自己決定的
你拿這些評價可以定義出一個 Loss
有了這些評價以後拿去訓練你的 Actor 
然後你可以更新你的參數一次 

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/f6393c61-65ac-478e-a7b2-4c634b146cd1)

但是你只能更新一次而已
一旦更新完一次參數以後 接下來你就要重新去蒐集資料了
這就是為什麼 RL 往往它的訓練過程非常花時間的原因

你每次 Update 完一次參數以後你的資料就要重新再蒐集一次
如果你參數要 Update 400 次 那你資料就要蒐集 400 次

Q:為什麼我們不能夠一組資料 就拿來 Update 模型 Update 400 次?
A:一個人的藥物 可能是另外一個人的毒藥

這些資料是由 θi-1 所蒐集出來的
這是 θi-1 跟環境互動的結果->這個是 θi-1 的經驗
這些經驗可以拿來更新 θi-1 可以拿來 Update θi-1 的參數
但它不一定適合拿來 Update θi 的參數

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1d324a24-9ac1-40b4-b53b-5585184395e8)

同一個 Action 同一個行為對於不同的 Actor 而言 它的好是不一樣的

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/3bb54c1a-cf04-4e18-a222-49c1f29c615d)

______________

要被訓練的 Actor 跟要拿來跟環境互動的 Actor最好是同一個
當我們訓練的 Actor 跟互動的 Actor 是同一個的時候 = On-policy (Off-policy 的 Learning 不會細講)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/354c05de-4a08-4663-bae3-538c93628494)

Off-policy 的 Learning 它可以想辦法讓 θi 去根據 θi-1 所蒐集的資料來進行學習
Off-policy的好處就是 你就不用一直蒐集資料
Proximal Policy Optimization - 非常經典的 Off-policy 的方法(PPO)

Off-policy 的重點就是 你在訓練的那個 Network 要知道自己跟別人之間的差距
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/29200f40-ba36-48f6-ad25-306d9446c41f)
就好像是你去問克里斯伊凡 (美國隊長) 怎麼追一個女生

________

### Exploration 

Actor 它在採取行為的時候它是有一些隨機性
而這個隨機性其實非常地重要
很多時候你隨機性不夠 你會 Train 不起來

假設你一開始初始的 Actor 它永遠都只會向右移動 它從來都不會知道要開火
如果它從來沒有採取開火這個行為 你就永遠不知道開火這件事情 到底是好還是不好
唯有今天某一個 Actor 去試圖做開火這件事得到 Reward 你才有辦法去評估這個行為好或不好
假設有一些 Action 從來沒被執行過 那你根本就無從知道 這個 Action 好或不好

所以你在 Training 的時候 你甚至會刻意加大它的隨機性 -> Exploration

比如說 有人會刻意加大 Actor 的 Output Distribution 的 Entropy
讓它在訓練的時候比較容易 Sample 到那些機率比較低的行為

比如說有人會直接在這個 Actor 它的那個參數上面加 Noise 
直接在 Actor 參數上加 Noise讓它每一次採取的行為都不一樣 
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/03bf4c20-3851-4355-b4da-a269b6437391)

那 Exploration 其實也是 RL Training 的過程中 一個非常重要的技巧
如果你在訓練過程中你沒有讓 Network 盡量去試不同的 Action 你很有可能你會 Train 不出好的結果
